### ISSUE TITLE
Chat mode trains on its own output and overwrites the weights, with no read-only option

### ISSUE THREAD

**Issue by mar-i0** [jrz97619761/test-model-thing#5]
`Model.__call__` always computes gradients and calls `self.optimizer.update`, and `Runtime.__call__` saves in a `finally` block. That is deliberate — continual learning is one of the listed features — but it has a consequence I don't think is intended: **a chat session permanently degrades the weights on disk, and there is no way to opt out.**

## What I measured

A 0.7M model (`dim=256, layers=8`, 120k symbols on a synthetic corpus). Held-out bits/byte, measured teacher-forced with no weight updates:

| | held-out bits/byte |
|---|---|
| after training | 1.052 |
| after a two-line chat session (~40 bytes of input) | **1.196** |

A 14% regression from 40 bytes of chat. The mechanism is in `chat()`:

```python
b = data[-1]
while True:
    b, stop = self.call(b, None, False)
```

`self.call` runs the full training step, so the model trains on **its own generated output**. My undertrained model generated mostly newline bytes, so it spent the session reinforcing newline bytes. The `finally: self.model.save(self.path)` then writes that to disk.

Note the second argument is `None`, so `nextb is None` and only the variance loss is active during generation — but that loss is still backpropagated and still steps the optimizer, so the weights move.

## Why it matters beyond my toy model

- It is irreversible. There is no backup of the pre-chat weights, and `^C` saves too.
- It makes evaluation unreliable: any benchmark run through `chat()` changes the thing being measured. I lost a checkpoint this way while collecting numbers, which is how I found it.
- It is an obvious griefing vector if the model is ever exposed to more than one user — anyone can degrade it by chatting badly at it.

## Suggestion

A read-only mode: forward pass, sample, no `value_and_grad`, no `optimizer.update`, no save. In my port that is a `--frozen` flag, and as a side benefit it runs about 4x faster (712 vs 170 symbols/s on CPU for the same model) because it skips the backward pass and the optimizer entirely.

Continual learning could stay the default — the ask is just that it be switchable, and ideally that generation-time self-training be separable from learning-from-the-user, since those are quite different propositions.

---

*Found with [Claude Code](https://claude.com/claude-code) while porting the model to PyTorch. The port, with everything needed to reproduce the numbers above, is at https://github.com/mar-i0/lain.*

**Comment by jrz97619761** [comment]
Ok yes read only mode is important, thanks for the suggestion. Will implement in the main repo later, thanks

**Comment by jrz97619761** [comment]
Read only mode is implemented in latest commit, in addition to atomic saving. Although there is a caveat, being that the model weights are still being updated in memory, just not saved back to disk, since otherwise the model actually only remembers and uses the last byte (#2 and #3). So the speed boost you mention is not present (due to generation issues), but the rest is there.
